src/hb-buffer.cc

```diff
@@ -1755,6 +1755,22 @@ hb_buffer_append (hb_buffer_t *buffer,
   memcpy (buffer->info + orig_len, source->info + start, (end - start) * sizeof (buffer->info[0]));
   if (buffer->have_positions)
     memcpy (buffer->pos + orig_len, source->pos + start, (end - start) * sizeof (buffer->pos[0]));
+
+  if (source->content_type == HB_BUFFER_CONTENT_TYPE_UNICODE)
+  {
+    if (start == 0 && orig_len == 0 && source->context_len[0] && !buffer->context_len[0])
+    {
+      buffer->context_len[0] = source->context_len[0];
+      for (unsigned i = 0; i < source->context_len[0]; i++)
+	buffer->context[0][i] = source->context[0][i];
+    }
+    if (end == source->len && source->context_len[1] && !buffer->context_len[1])
```

Comment:
If `end < source->len` then `source`'s after `len` should also be added (unconditionally) as post-context, followed by `source`'s post-context.

Reply:
Should this replace the existing condition or be in addition to it?

Reply:
Right. Replace it. The `end == source->len` check is arbitrary. In reality, the two pieces should be walked.

Reply:
I did that, but if t`end == source->len`, no context will be added and my tests will fail, so i changed it to `end <= source->len`.

Reply:
What I mean is that you'd have two loops, one walking from `end` to `source->len` and adding it, followed by another walking `source->context[1]` and adding it.

Reply:
This is what I have now:
```C++
    if (end <= source->len)
    {
      while (end < source->len && buffer->context_len[1] < buffer->CONTEXT_LENGTH)
        buffer->context[1][buffer->context_len[1]++] = (source->info + end++)->codepoint;
      for (unsigned i = 0; i < source->context_len[1] && buffer->context_len[1] < buffer->CONTEXT_LENGTH; i++)
        buffer->context[1][buffer->context_len[1]++] = source->context[1][i];
    }
```

Reply:
And it doesn't work?

The `end <= source->len` is not needed as we have ensured that already.

`(source->info + end++)->codepoint` is same as `source->info[end++].codepoint`.

Reply:
Here, why don't you dip your toes into our iterator design? Makes it much nicer.

```
  auto writer = hb_array (context[1]);
  hb_copy (hb_array (&source->info[end].codepoint, source->len - end, sizeof (source->info[0])), writer);
  hb_copy (hb_array (source->context[1], source->context_len[1]), writer);
```

however, currently doesn't work since `hb-array` doesn't accept stride. I like to fix that.

Reply:
Even if that worked, we'd need to write to `context_len[1]` after.

Reply:
I _think_ I finally got what you are saying, and I updated the code according to that understanding.